lib/src/alloc.h

```diff
@@ -9,6 +9,21 @@ extern "C" {
 #include <stdbool.h>
 #include <stdio.h>
 
+// Allow clients to override allocation functions
+
+#ifndef ts_malloc
+#define ts_malloc(size)         ts_malloc_default(size)
+#endif
+#ifndef ts_calloc
+#define ts_calloc(count,size)   ts_calloc_default(count,size)
+#endif
+#ifndef ts_realloc
+#define ts_realloc(buffer,size) ts_realloc_default(buffer,size)
+#endif
+#ifndef ts_free
+#define ts_free(buffer)         ts_free_default(buffer)
+#endif
```

Comment:
Sorry, my earlier feedback was a little bit vague. In that comment I made about the `TREE_SITTER_TEST` case, I actually meant something slightly different. I don't think we should let people override `ts_malloc` etc when running the tests; I just thought you could make the header a little bit simpler.

I was thinking something like this:

```c
#if defined(TREE_SITTER_TEST)

void *ts_record_malloc(size_t);
void *ts_record_calloc(size_t, size_t);
void *ts_record_realloc(void *, size_t);
void ts_record_free(void *);
bool ts_toggle_allocation_recording(bool);

#define ts_malloc  ts_record_malloc
#define ts_calloc  ts_record_calloc
#define ts_realloc ts_record_realloc
#define ts_free    ts_record_free

// ...

#endif
```

Then, below, (and outside of the previous `#if` condition) we could provide the default definitions:

```c
#ifndef ts_malloc
#define ts_malloc(size)         ts_malloc_default(size)
#endif
#ifndef ts_calloc
#define ts_calloc(count,size)   ts_calloc_default(count,size)
#endif
#ifndef ts_realloc
#define ts_realloc(buffer,size) ts_realloc_default(buffer,size)
#endif
#ifndef ts_free
#define ts_free(buffer)         ts_free_default(buffer)
#endif
```

Does that seem ok to you?

Reply:
Also, a very minor question: For brevity, can we just define the default macros as plain name-substitutions? For example:

```c
#ifndef ts_malloc
#define ts_malloc ts_malloc_default
#endif
```

I think that since `malloc` and friends are standard functions, it's still readable and clear what the intended arguments are.

Reply:
sure, np! that makes more sense since I also thought it odd to allow user overrides in tests

re: #2 - _[tiny nitpick]_  imho, it's better to keep the args in the macros for clarity since the order of parameters changes based on `{platform} x {C,C++} x {langversion}`
granted, it's only for aligned allocations so not applicable for ones above hence the _[tiny nitpick]_

thoughts?

Reply:
I guess on that issue, I'm ok with either way. But I think for these 4 standard allocation APIs, `malloc`, `calloc`, `realloc` and `free`, the arguments are pretty clear, and it reads a little more clearly without the argument lists.

Reply:
I'll keep it without the args then since i think consistent style >> nitpick

Swamped with work stuff so will fix it up in a couple of day